### PR TITLE
Feature/form templates json validation

### PR DESCRIPTION
**Pull request by jeberhardt**
# Summary | Résumé

This will validate the JSON that is being uploaded for a form.

# Test instructions | Instructions pour tester la modification

Launch application and upload a form. You can test for a form failing by removing any of the top-level elements: "form", "submission", "internalTitleEn", "internalTitleFr", "publishingStatus". The form will not be uploaded, and an error will appear in the Javascript console in the browser.

# Unresolved questions / Out of scope | Questions non résolues ou hors sujet

>> The schema will need to be updated to be more accurate, describing the required / optional properties, and the data structure.
>> The error will need to be presented properly to the user in the UI.

### CODE REVIEW

**Review by Moro-Code — CHANGES_REQUESTED** [review]
Code looks fantastic ! Awesome work but we need unit tests on the middleware you created at a minimum

**Review by craigzour — APPROVED** [review]

File: tests/templates.test.js [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+jest.mock("next-auth/client");
+
+jest.mock("@aws-sdk/client-lambda", () => {
+  return {
+    LambdaClient: jest.fn(() => {
+      return {
+        send: jest.fn((command) => {
+          const encoder = new TextEncoder(),
+            decoder = new TextDecoder();
```
Comment (Moro-Code) [inline tests/templates.test.js:20]: If these are going to be on separate lines anyways. Might as well declare them in separate lines. I was a little confused reading this

File: tests/templates.test.js [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        Origin: "http://localhost:3000",
+      },
+      body: JSON.stringify({
+        formConfig: brokenFormTemplate,
+        method: "INSERT",
+      }),
+    });
+
+    await templates(req, res);
+    expect(res.statusCode).toBe(400);
```
Comment (Moro-Code) [inline tests/templates.test.js:116]: Something else to check is if you're getting the error messages you expect in the response

File: tests/templates.test.js [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        }),
+      };
+    }),
+    InvokeCommand: jest.fn((payload) => {
+      return payload;
+    }),
+  };
+});
+
+describe("Test JSON validation scenarios", () => {
```
Comment (Moro-Code) [inline tests/templates.test.js:64]: Overall these tests look good ! Well done ! There are some additional test cases like for example the lambda client throwing an error but this is outside the scope of this PR

File: lib/jsonValidator/jsonValidator.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
+import { Schema, Validator } from "jsonschema";
+import { NextApiRequest, NextApiResponse } from "next";
+
+export type validateOptions = {
+  jsonKey: string;
+};
+
+const validate = (
```
Comment (Moro-Code) [inline lib/jsonValidator/jsonValidator.ts:8]: I still need unit tests for this. You wrote unit tests for the template handler but this doesn't provide adequate coverage for this file. We need unit tests specifically for the validate middleware

File: lib/http/allowed.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
+import { NextApiRequest, NextApiResponse } from "next";
+import { getSession } from "next-auth/client";
+
+const isRequestAllowed = (
```
Comment (Moro-Code) [inline lib/http/allowed.ts:4]: I still need unit tests for this. You wrote unit tests for the template handler but this doesn't provide adequate coverage for this file. We need unit tests specifically for the isRequestAllowed middleware

File: lib/jsonValidator/jsonValidator.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
+import { Schema, Validator } from "jsonschema";
+import { NextApiRequest, NextApiResponse } from "next";
+
+export type validateOptions = {
```
Comment (Moro-Code) [inline lib/jsonValidator/jsonValidator.ts:4]: This should be capitalized if its a type

**Review by Moro-Code — CHANGES_REQUESTED** [review]
Great work adding tests for the API handler itself. This will certainly help the team to increase test coverage of the code base. You're still missing unit tests for the individual middleware functions. I also left some minor comments

**Review by falila — APPROVED** [review]
Great work @jeberhardt. 
A small detail which maybe out of the scope of this PR. See pic. 
IMO , The red circle's content should be removed whenever there is an error otherwise  users might be confused. 
Screen Shot 2021-10-18 at 1 38 21 PM
